src/coreclr/src/jit/optimizer.cpp

```diff
@@ -8659,6 +8659,145 @@ bool Compiler::optIsRangeCheckRemovable(GenTree* tree)
+            // both blocks are expected to be BBJ_RETURN
+            // TODO: implement for GT_ASG
+            (trueBb->bbJumpKind != BBJ_RETURN) || (falseBb->bbJumpKind != BBJ_RETURN) ||
+            // give up if these blocks are used by someone else
+            (trueBb->countOfInEdges() != 1) || (falseBb->countOfInEdges() != 1) ||
+            (trueBb->bbFlags & BBF_DONT_REMOVE) || (falseBb->bbFlags & BBF_DONT_REMOVE))
+        {
+            continue;
+        }
+
+        Statement* trueBbStmt  = trueBb->firstStmt();
+        Statement* falseBbStmt = falseBb->firstStmt();
+        if ((trueBbStmt == nullptr) || (trueBbStmt->GetPrevStmt() != trueBbStmt) ||
+            // make sure both blocks are single statement
+            (falseBbStmt == nullptr) || (falseBbStmt->GetPrevStmt() != falseBbStmt))
+        {
+            continue;
+        }
+
+        // make sure both blocks are single `return cns` nodes
+        GenTree* retTrueNode  = trueBbStmt->GetRootNode();
+        GenTree* retFalseNode = falseBbStmt->GetRootNode();
+        if (!retTrueNode->OperIs(GT_RETURN) || !retFalseNode->OperIs(GT_RETURN) ||
+            // both blocks of the same types
+            !retTrueNode->TypeIs(typ) || !retFalseNode->TypeIs(typ) ||
+            // both blocks are `return cns`
+            !retTrueNode->gtGetOp1()->IsIntegralConst() || !retFalseNode->gtGetOp1()->IsIntegralConst())
+        {
+            continue;
+        }
+
+        // TODO: optimize  `condition ? x + 2 : x + 3`
+        // same LCL_VAR in both blocks
+
+        ssize_t cnsForTrue  = retTrueNode->gtGetOp1()->AsIntCon()->IconValue();
+        ssize_t cnsForFalse = retFalseNode->gtGetOp1()->AsIntCon()->IconValue();
+
+        // the optimization can be applied when the difference between two constants is 1
+        // because it's what `cmp\test` return for true (and 0 for false)
+        if (abs(cnsForTrue - cnsForFalse) != 1)
+        {
+            // TODO: for some bigger difference we can use GT_LEA (?) or even cmov
+            continue;
+        }
+
+        ssize_t addValue;
+        if (cnsForFalse > cnsForTrue)
+        {
+            addValue = cnsForTrue;
+            // cns for true-branch must be greater that cns for false-branch
+            // we need to flip the comparision operator in this case, e.g. `>` -> `<=`
+            assert(rootSubNode->OperIsCompare());
+            gtReverseCond(rootSubNode);
+        }
+        else
+        {
+            addValue = cnsForFalse;
+        }
+
+        // unlink both blocks
+        fgRemoveRefPred(block->bbJumpDest, block);
+        fgRemoveRefPred(block->bbNext, block);
+
+        // convert current block from BBJ_COND to BBJ_RETURN
+        block->bbJumpKind = BBJ_RETURN;
+        rootNode->ChangeOperUnchecked(GT_RETURN);
+        rootNode->AsOp()->gtOp1 = gtNewOperNode(GT_ADD, typ, rootSubNode, gtNewIconNode(addValue, typ));
+        rootNode->ChangeType(typ);
```

Comment:
>rootNode->ChangeType(typ); [](start = 8, length = 26)

wondering why we would this need. Won't `rootNode`'s type will be `typ` at this point?

Reply:
it's `TYP_VOID` because it used to be `GT_JTRUE`